AI #1578: Enhance definition of attributes (#1960)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/attributes_overview.md

Reviewer: The phrase “requirements defined in attributes” contradicts the PR purpose statement that Attributes are not first-class entities that receive requirements.

Am I missing something?

Author: @davidearney, Good catch, you are not missing anything. The first sentence needs better wording.

### Explanation
Attributes do not "receive" requirements as independent entities. An Attribute is a reusable container that defines a set of requirements to be applied to whatever dataset, column, or object references (conforms to) that Attribute.
To avoid confusion, we can tighten the wording to something like:

### Proposal
>"Attributes define reusable sets of requirements that apply to specification entities (datasets, columns, objects, etc.), rather than being requirements on the Attribute itself."

Will that help?

# Attributes

> Attributes define reusable sets of requirements that apply either across a [*FOCUS dataset*](#glossary:FOCUS-dataset) or an individual column level. An attribute is associated with a dataset or column in their respective Requirements sections to indicate that the dataset or column conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the relevant dataset, column, or data structure being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.

Reviewer 2: ```suggestion
Attributes define reusable sets of requirements that apply either across a [*FOCUS dataset*](#glossary:FOCUS-dataset) or an individual column level. An attribute is associated with a dataset or column in their respective Requirements sections to indicate that the specification entity (dataset, column, etc) conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the relevant dataset, column, or data structure being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.
```

> Attributes define reusable sets of requirements that apply either across a [*FOCUS dataset*](#glossary:FOCUS-dataset) or other specification entity (column, object, etc.). An attribute is associated with a specification entity in their respective Requirements sections to indicate that the specification entity conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the specification entity being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.

Author: ```suggestion
Attributes define reusable sets of requirements that apply to specification entities (datasets, columns, objects, etc.), rather than being requirements on the Attribute itself. An attribute is associated with a specification entity in their respective Requirements sections to indicate that the specification entity conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the specification entity being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.
```

> Attributes define reusable sets of requirements that apply to specification entities (datasets, columns, objects, etc.), rather than being requirements on the Attribute itself. An attribute is associated with a specification entity in their respective Requirements sections to indicate that the specification entity conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the specification entity being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.

Reviewer 3: @davidearney @jpradocueva 
I understand the point y'all are trying to make (and solve) now, but putting that clarification in the very first sentence feels like it confuses the situation more.  Also, it doesn't scrub the "Requirements defined in attributes" phrase that @davidearney was concerned about.  AND, the original paragraph feels like it repeats itself a couple times.  How about the following?  I know it's a lot shorter, but I think it still sells all the points the original had.

```suggestion
Attributes serve as reusable containers for requirements that enforce consistency across FOCUS specification entities (e.g., datasets, columns, objects). Functioning as a governance layer, Attributes define the constraints -- such as naming conventions, data types, granularity, or recency -- that an entity must satisfy. By grouping these requirements, Attributes ensure that data from any origin can be processed via standard instructions, which is essential for accurately supporting [FinOps capabilities][FODOFC].
```

Author: My preference is to leave the list open with an "etc." at the end so we don't need to revisit if the group defines other entities: " ... such as datasets, columns, objects, etc."

Reviewer 3: @jpradocueva 
I have updated the suggestion to use an `e.g.,`, which we use often throughout the spec, and which is not meant to be an exhaustive list (unlike `i.e.,`).  Let me know if that works for you.

> Attributes define reusable sets of requirements that apply either across a [*FOCUS dataset*](#glossary:FOCUS-dataset) or an individual column level. An attribute is associated with a dataset or column in their respective Requirements sections to indicate that the specification entity (dataset, column, etc) conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the relevant dataset, column, or data structure being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.

Author: ```suggestion
Attributes define reusable sets of requirements that apply either across a [*FOCUS dataset*](#glossary:FOCUS-dataset) or other specification entity (column, object, etc.). An attribute is associated with a specification entity in their respective Requirements sections to indicate that the specification entity conforms to the attribute.  Requirements on data content can include naming conventions, data types, formatting standardizations, etc. Attributes may introduce high-level requirements for data granularity, recency, frequency, etc. Within an attribute, each requirement applies to the specification entity being constrained; the attribute serves as the grouping mechanism. Requirements defined in attributes are necessary for servicing [FinOps capabilities][FODOFC] accurately using a standard set of instructions regardless of the origin of the data.
```
